File: backend/app/tasks/scraping_tasks.py
import logging
import os
import csv
from datetime import datetime
from typing import Dict, List
from sqlalchemy.orm import Session

from app.tasks.celery_app import celery_app
from app.db.session import SessionLocal
from app.models import SearchJob, Paper
from app.services.scholar_scraper import GoogleScholarScraper
from app.services.semantic_filter import SemanticFilter
from app.services.email_service import EmailService
from app.core.config import settings

logger = logging.getLogger(__name__)


@celery_app.task(bind=True, max_retries=3)
def run_search_job(self, job_id: str):
    """
    Celery task to run a search job.

    This task:
    1. Retrieves the job from database
    2. Runs the Google Scholar scraper
    3. Applies semantic filtering if configured
    4. Saves results to database and CSV
    5. Sends email notification
    6. Handles errors and supports resumption

    Args:
        job_id: UUID of the SearchJob
    """
    db = SessionLocal()

    try:
        # Get job from database
        job = db.query(SearchJob).filter(SearchJob.id == job_id).first()
        if not job:
            logger.error(f"Job {job_id} not found")
            return

        # Update job status
        job.status = "running"
        job.status_message = "Initializing browser and starting search..."
        job.started_at = datetime.utcnow()
        job.celery_task_id = self.request.id
        db.commit()

        logger.info(f"Starting search job {job_id}")

        # Check if we're resuming from a checkpoint
        start_year = job.start_year
        if job.last_checkpoint and job.last_checkpoint.get('last_year_completed'):
            start_year = job.last_checkpoint['last_year_completed'] + 1
            job.status_message = f"Resuming from year {start_year}..."
            db.commit()
            logger.info(f"Resuming from year {start_year}")

        # Initialize scraper with screenshot support
        screenshot_dir = os.path.join(settings.UPLOAD_DIR, "screenshots")
        os.makedirs(screenshot_dir, exist_ok=True)

        scraper = GoogleScholarScraper(
            headless=True,
            job_id=str(job_id),
            screenshot_dir=screenshot_dir
        )

        # Track already saved paper titles to avoid duplicates
        saved_paper_titles = set()

        # Progress callback to update database and save papers incrementally
        def update_progress(current: int, estimated_total: int):
            try:
                progress = min((current / max(estimated_total, 1)) * 100, 100)
                job.progress = progress
                job.papers_processed = current
                job.status_message = f"Collecting papers... {current} found so far"
                db.commit()
                logger.info(f"Job {job_id}: {progress:.1f}% complete ({current} papers)")
            except Exception as e:
                logger.error(f"Error updating progress: {e}")

        # New callback to save papers incrementally
        def save_papers_callback(papers_batch: list):
            try:
                for paper_data in papers_batch:
                    title = paper_data.get('title', '')
                    # Skip if already saved
                    if title in saved_paper_titles:
                        continue

                    saved_paper_titles.add(title)
                    paper = Paper(
                        search_job_id=job.id,
                        title=title,
                        authors=paper_data.get('authors', ''),
                        year=paper_data.get('year'),
                        abstract=paper_data.get('abstract', ''),
                        source=paper_data.get('source', ''),
                        citations=paper_data.get('citations', 0),
                        url=paper_data.get('url', ''),
                        publisher=paper_data.get('publisher', ''),
                        semantic_score=paper_data.get('semantic_score')
                    )
                    db.add(paper)
                db.commit()
                logger.info(f"Saved {len(papers_batch)} papers to database")
            except Exception as e:
                logger.error(f"Error saving papers incrementally: {e}")
                db.rollback()

        # Run the search
        papers = scraper.search(
            keywords_include=job.keywords_include,
            keywords_exclude=job.keywords_exclude or [],
            start_year=start_year,
            end_year=job.end_year,
            progress_callback=update_progress,
            papers_callback=save_papers_callback
        )

        logger.info(f"Scraping complete: {len(papers)} papers found")
        job.status_message = f"Scraping complete. Found {len(papers)} papers."
        db.commit()

        # Apply semantic filtering if configured
        if job.semantic_criteria:
            logger.info("Applying semantic filtering...")
            job.status_message = "Applying semantic filtering to papers..."
            db.commit()
            semantic_filter = SemanticFilter()

            inclusion_criteria = job.semantic_criteria.get('inclusion')
            exclusion_criteria = job.semantic_criteria.get('exclusion')

            papers = semantic_filter.filter_papers(
                papers,
                inclusion_criteria=inclusion_criteria,
                exclusion_criteria=exclusion_criteria
            )

            logger.info(f"After semantic filtering: {len(papers)} papers remain")
            job.status_message = f"Semantic filtering complete. {len(papers)} papers passed criteria."
            db.commit()

        # Save any remaining papers to database (ones not saved incrementally due to semantic filtering)
        logger.info("Saving any remaining papers to database...")
        job.status_message = "Saving filtered papers to database..."
        db.commit()
        papers_saved = 0
        for paper_data in papers:
            title = paper_data.get('title', '')
            # Skip if already saved incrementally
            if title in saved_paper_titles:
                continue

            saved_paper_titles.add(title)
            paper = Paper(
                search_job_id=job.id,
                title=title,
                authors=paper_data.get('authors', ''),
                year=paper_data.get('year'),
                abstract=paper_data.get('abstract', ''),
                source=paper_data.get('source', ''),
                citations=paper_data.get('citations', 0),
                url=paper_data.get('url', ''),
                publisher=paper_data.get('publisher', ''),
                semantic_score=paper_data.get('semantic_score')
            )
            db.add(paper)
            papers_saved += 1

        if papers_saved > 0:
            logger.info(f"Saved {papers_saved} additional papers after filtering")

        # Export to CSV
        csv_filename = f"search_{job_id}_{datetime.now().strftime('%Y%m%d_%H%M%S')}.csv"
        csv_path = os.path.join(settings.UPLOAD_DIR, csv_filename)

        os.makedirs(settings.UPLOAD_DIR, exist_ok=True)

        logger.info(f"Exporting to CSV: {csv_path}")
        job.status_message = "Exporting results to CSV file..."
        db.commit()
        _export_to_csv(papers, csv_path)

        # Update job
        job.status = "completed"
        job.status_message = f"Job completed successfully! {len(papers)} papers found."
        job.completed_at = datetime.utcnow()
        job.total_papers_found = len(papers)
        job.papers_processed = len(papers)
        job.progress = 100.0
        job.csv_file_path = csv_path
        job.last_checkpoint = None  # Clear checkpoint on success
        db.commit()

        logger.info(f"Job {job_id} completed successfully")

        # Send email notification
        try:
            email_service = EmailService()
            user = job.user

            download_url = f"{settings.FRONTEND_URL}/download/{job_id}"

            email_service.send_job_completion_email(
                to_email=user.email,
                user_name=user.name or "User",
                job_name=job.name,
                total_papers=len(papers),
                download_url=download_url
            )

            logger.info(f"Email sent to {user.email}")

        except Exception as e:
            logger.error(f"Error sending email: {e}")
            # Don't fail the job if email fails

    except Exception as e:
        logger.error(f"Error in search job {job_id}: {e}", exc_info=True)

        # Update job with error
        job.status = "failed"
        job.error_message = str(e)
        job.status_message = f"Error: {str(e)}"
        job.retry_count += 1

        # Save checkpoint for resumption
        if job.end_year and job.start_year:
            # Estimate last completed year based on progress
            if job.progress > 0:
                year_range = job.end_year - job.start_year + 1
                years_completed = int((job.progress / 100.0) * year_range)
                last_year = job.start_year + years_completed - 1

                job.last_checkpoint = {
                    'last_year_completed': last_year,
                    'papers_collected': job.papers_processed,
                    'timestamp': datetime.utcnow().isoformat()
                }

        db.commit()

        # Retry if we haven't exceeded max retries
        if job.retry_count < settings.SCRAPER_RETRY_ATTEMPTS:
            logger.info(f"Retrying job {job_id} (attempt {job.retry_count})")
            raise self.retry(exc=e, countdown=settings.SCRAPER_RETRY_DELAY * job.retry_count)
        else:
            logger.error(f"Job {job_id} failed after {job.retry_count} retries")

    finally:
        db.close()


def _export_to_csv(papers: List[Dict], file_path: str):
    """
    Export papers to CSV file.

    Args:
        papers: List of paper dictionaries
        file_path: Path to save CSV file
    """
    if not papers:
        # Create empty file with headers
        with open(file_path, 'w', newline='', encoding='utf-8') as f:
            writer = csv.writer(f)
            writer.writerow([
                'Title', 'Authors', 'Year', 'Source', 'Publisher',
                'Citations', 'Abstract', 'URL'
            ])
        return

    fieldnames = [
        'title', 'authors', 'year', 'source', 'publisher',
        'citations', 'abstract', 'url'
    ]

    with open(file_path, 'w', newline='', encoding='utf-8') as f:
        writer = csv.DictWriter(f, fieldnames=fieldnames, extrasaction='ignore')

        # Write header with nice names
        writer.writerow({
            'title': 'Title',
            'authors': 'Authors',
            'year': 'Year',
            'source': 'Source',
            'publisher': 'Publisher',
            'citations': 'Citations',
            'abstract': 'Abstract',
            'url': 'URL'
        })

        # Write data
        for paper in papers:
            writer.writerow(paper)


@celery_app.task
def resume_failed_job(job_id: str):
    """
    Resume a failed job from its last checkpoint.

    Args:
        job_id: UUID of the SearchJob to resume
    """
    db = SessionLocal()

    try:
        job = db.query(SearchJob).filter(SearchJob.id == job_id).first()
        if not job:
            logger.error(f"Job {job_id} not found")
            return

        if job.status != "failed":
            logger.warning(f"Job {job_id} is not in failed state")
            return

        # Reset job status
        job.status = "pending"
        job.error_message = None
        db.commit()

        # Start the job again (it will resume from checkpoint)
        run_search_job.delay(job_id)

        logger.info(f"Resumed job {job_id}")

    finally:
        db.close()

File: backend-node/src/tasks/scrapingTasks.ts
/**
 * Comprehensive Scraping Task Orchestration
 * Uses multi-strategy scraper, semantic filtering, and generates all outputs
 */
import { SearchJob, Paper, User } from '../models';
import { MultiStrategyScholarScraper } from '../services/scrapers/multiStrategyScraper';
import SemanticFilter from '../services/semanticFilter';
import emailService from '../services/emailService';
import { writePapersToCSV } from '../services/csvWriter';
import { generatePrismaDiagram } from '../services/prismaDiagram';
import { generateBibtexFile } from '../services/bibtexGenerator';
import { generateLatexReview } from '../services/latexGenerator';
import logger from '../core/logger';
import { config } from '../core/config';
import path from 'path';

export async function runSearchJob(jobId: string): Promise<void> {
  logger.info(`Running search job: ${jobId}`);

  const scraper = new MultiStrategyScholarScraper({
    useTor: config.USE_TOR, // Use Tor if enabled in .env
    headless: true,
    jobId,
    screenshotDir: path.join(config.UPLOAD_DIR, 'screenshots'),
  });

  try {
    const job = await SearchJob.findByPk(jobId);

    if (!job) {
      throw new Error(`Job ${jobId} not found`);
    }

    // Update job status
    job.status = 'running';
    job.startedAt = new Date();
    job.statusMessage = 'Initializing multi-strategy scraper...';
    await job.save();

    // Progress callback
    const progressCallback = async (current: number, total: number) => {
      const progress = (current / (total || 1)) * 50; // First 50% is scraping
      job.progress = Math.min(progress, 50);
      job.totalPapersFound = current;
      await job.save();
    };

    // Incremental paper save callback
    const papersCallback = async (newPapers: any[]) => {
      for (const paper of newPapers) {
        await Paper.create({
          searchJobId: jobId,
          title: paper.title,
          authors: paper.authors,
          year: paper.year,
          source: paper.source,
          publisher: paper.publisher,
          citations: paper.citations || 0,
          abstract: paper.abstract,
          url: paper.url,
        });
      }
      job.papersProcessed += newPapers.length;
      await job.save();
    };

    job.statusMessage = 'Searching Google Scholar with multi-strategy approach...';
    await job.save();

    // Perform search with multi-strategy scraper
    const papers = await scraper.search({
      keywords: job.keywordsInclude,
      startYear: job.startYear || undefined,
      endYear: job.endYear || undefined,
      maxResults: job.maxResults || undefined,
      progressCallback,
      papersCallback,
    });

    logger.info(`Found ${papers.length} papers for job ${jobId}`);

    // Log strategy stats
    const strategyStats = scraper.getStrategyStats();
    logger.info(`Strategy stats: ${JSON.stringify(strategyStats)}`);

    job.totalPapersFound = papers.length;
    job.progress = 50;
    job.statusMessage = `Found ${papers.length} papers. Processing...`;
    await job.save();

    // Get all papers from database
    let savedPapers = await Paper.findAll({
      where: { searchJobId: jobId },
    });

    // Apply semantic filter if criteria provided
    let includedPapers = savedPapers;
    if (job.semanticCriteria) {
      job.statusMessage = 'Applying semantic filtering with Gemini AI...';
      job.progress = 60;
      await job.save();

      const filter = new SemanticFilter();
      const papersForFilter = savedPapers.map((p) => ({
        title: p.title,
        abstract: p.abstract || '',
      }));

      const filtered = await filter.filterPapers(
        papersForFilter,
        job.semanticCriteria.inclusion,
        job.semanticCriteria.exclusion,
        job.semanticBatchMode ? 10 : 1
      );

      // Update semantic scores and exclusion status for ALL papers
      for (const filteredPaper of filtered) {
        const paper = savedPapers.find((p) => p.title === filteredPaper.title);
        if (paper) {
          paper.semanticScore = filteredPaper.semantic_score || 0;
          paper.isExcluded = filteredPaper.is_excluded || false;
          paper.exclusionReason = filteredPaper.exclusion_reason || undefined;
          paper.semanticRationale = filteredPaper.semantic_rationale || undefined;
          await paper.save();
        }
      }

      includedPapers = savedPapers.filter((p) => !p.isExcluded);
      logger.info(`Semantic filtering: ${includedPapers.length}/${savedPapers.length} papers included, ${savedPapers.length - includedPapers.length} excluded`);
    }

    job.progress = 70;
    job.statusMessage = 'Generating outputs...';
    await job.save();

    // Generate CSV with ALL papers (including excluded ones)
    logger.info('Generating CSV export...');
    const csvPath = await writePapersToCSV(
      savedPapers.map((p) => ({
        title: p.title,
        authors: p.authors || '',
        year: p.year,
        source: p.source || '',
        publisher: p.publisher || '',
        citations: p.citations,
        abstract: p.abstract || '',
        url: p.url || '',
        doi: p.doi || '',
        semanticScore: p.semanticScore,
        isExcluded: p.isExcluded,
        exclusionReason: p.exclusionReason,
        semanticRationale: p.semanticRationale,
      })),
      jobId,
      job.name
    );
    job.csvFilePath = csvPath;
    job.progress = 75;
    await job.save();

    // Generate PRISMA metrics (following PRISMA 2020 standards)
    const duplicatesRemoved = papers.length - savedPapers.length; // Papers found - papers saved (duplicates)
    const semanticExcluded = savedPapers.length - includedPapers.length;

    const prismaMetrics = {
      identification: {
        records_identified: papers.length,
      },
      screening: {
        records_excluded_duplicates: duplicatesRemoved,
        records_after_duplicates_removed: savedPapers.length,
      },
      eligibility: {
        full_text_assessed: savedPapers.length,
        full_text_excluded_semantic: semanticExcluded,
      },
      included: {
        studies_included: includedPapers.length,
      },
    };
    job.prismaMetrics = prismaMetrics;

    // Generate PRISMA diagram
    logger.info('Generating PRISMA diagram...');
    const prismaDiagramPath = generatePrismaDiagram(prismaMetrics, jobId, job.name);
    job.prismaDiagramPath = prismaDiagramPath;
    job.progress = 80;
    await job.save();

    // Generate BibTeX file (only included papers)
    logger.info('Generating BibTeX file...');
    const bibtexFilename = `${job.name.replace(/\s+/g, '_')}_References_${jobId}.bib`;
    const bibtexPath = path.join(config.UPLOAD_DIR, bibtexFilename);
    generateBibtexFile(
      includedPapers.map((p) => ({
        title: p.title,
        authors: p.authors,
        year: p.year,
        source: p.source,
        publisher: p.publisher,
        abstract: p.abstract,
        url: p.url,
      })),
      bibtexPath
    );
    job.bibtexFilePath = bibtexPath;
    job.progress = 85;
    await job.save();

    // Generate LaTeX document if requested
    if (job.generateLatex) {
      logger.info('Generating LaTeX systematic review...');
      job.statusMessage = 'Generating AI-powered LaTeX document...';
      await job.save();

      const latexFilename = `${job.name.replace(/\s+/g, '_')}_Review_${jobId}.tex`;
      const latexPath = path.join(config.UPLOAD_DIR, latexFilename);

      await generateLatexReview({
        papers: includedPapers.map((p) => ({
          title: p.title,
          authors: p.authors,
          year: p.year,
          source: p.source,
          publisher: p.publisher,
          abstract: p.abstract,
          url: p.url,
          citations: p.citations,
        })),
        searchCriteria: {
          keywords_include: job.keywordsInclude,
          keywords_exclude: job.keywordsExclude,
          start_year: job.startYear,
          end_year: job.endYear,
        },
        prismaMetrics,
        title: `Systematic Literature Review: ${job.name}`,
        outputPath: latexPath,
      });

      job.latexFilePath = latexPath;
      job.progress = 95;
      await job.save();
    }

    // Mark as completed
    job.progress = 100;
    job.status = 'completed';
    job.completedAt = new Date();
    job.statusMessage = `Completed! Found ${savedPapers.length} papers (${includedPapers.length} included, ${savedPapers.length - includedPapers.length} excluded).`;
    await job.save();

    logger.info(`Search job ${jobId} completed successfully`);

    // Send completion email
    const user = await User.findByPk(job.userId);
    if (user) {
      try {
        const downloadUrl = `${config.FRONTEND_URL}/jobs/${jobId}`;
        await emailService.sendJobCompletionEmail(
          user.email,
          user.name || 'User',
          job.name,
          includedPapers.length,
          downloadUrl
        );
      } catch (emailError) {
        logger.error('Failed to send completion email:', emailError);
        // Don't fail the job if email fails
      }
    }
  } catch (error) {
    logger.error(`Error running search job ${jobId}:`, error);

    const job = await SearchJob.findByPk(jobId);
    if (job) {
      job.status = 'failed';
      job.errorMessage = error instanceof Error ? error.message : 'Unknown error';
      job.statusMessage = 'Job failed';
      await job.save();
    }

    throw error;
  }
}

export async function resumeSearchJob(jobId: string): Promise<void> {
  logger.info(`Resuming search job: ${jobId}`);

  try {
    const job = await SearchJob.findByPk(jobId);
    if (!job) {
      throw new Error(`Job ${jobId} not found`);
    }

    // For now, just restart the job
    // TODO: Implement proper checkpoint-based resumption using lastCheckpoint
    job.status = 'pending';
    job.errorMessage = undefined;
    job.progress = 0;
    await job.save();

    await runSearchJob(jobId);
  } catch (error) {
    logger.error(`Error resuming search job ${jobId}:`, error);

    const job = await SearchJob.findByPk(jobId);
    if (job) {
      job.status = 'failed';
      job.errorMessage = error instanceof Error ? error.message : 'Unknown error';
      await job.save();
    }

    throw error;
  }
}
